lua/dap.lua

```diff
@@ -343,6 +328,52 @@ function Session.event_output(_, body)
   repl.append(body.output, '$')
 end
 
+function Session:_current_frame_index()
+  local current_frame_index = nil
+  if self.stopped_thread_id and self.threads[self.stopped_thread_id].frames then
+    local frames = self.threads[self.stopped_thread_id].frames
+    for i, f in ipairs(frames) do
+      if self.current_frame.id == f.id then
+        current_frame_index = i
+      end
+    end
+  end
+  return current_frame_index
+end
+
+function Session:_request_scopes(current_frame)
+  self:request('scopes', { frameId = current_frame.id }, function(_, scopes_resp)
+    if not scopes_resp or not scopes_resp.scopes then return end
+
+    current_frame.scopes = {}
+    for _, scope in pairs(scopes_resp.scopes) do
+
+      table.insert(current_frame.scopes, scope)
+      if not scope.expensive then
+        self:request('variables', { variablesReference = scope.variablesReference }, function(_, variables_resp)
+          if not variables_resp then return end
+
+          scope.variables = variables_resp.variables
+        end)
+      end
+    end
+  end)
+end
+
+function Session:_frame_delta(delta)
+  if self.stopped_thread_id and self.threads[self.stopped_thread_id].frames then
+    local frames = self.threads[self.stopped_thread_id].frames
+    local current_frame_index = self:_current_frame_index()
```

Comment:
How about something like this:

```
local function index_of(items, predicate)
  for i, item in ipairs(items) do
    if predicate(item) then
      return i
    end
  end
  return nil
end
```

```suggestion
    local current_frame_index = index_of(frames, function(f) return f.id == current end)
    assert(current_frame_index, 'current_frame must exist within frames')
```

The `_current_frame_index` method could then be deleted. It's only used once and a mix of abstract/concrete.
`index_of` would be a more generic concept (that people may be familiar with from other languages standard libraries) - it would also avoid re-retrieving the `frames` again which are already in scope here.